### DISCUSSION TITLE
Grid-Stat config file interp setting

### DISCUSSION THREAD

**Discussion by jwraby** [dtcenter/METplus#2105] [Configuration] [answered]
Is it possible to set the interp to "NONE" to prevent smoothing? The User's Guide provides the following possible settings:
FCST,OBS,BOTH.
If one cannot use "NONE" is there a reason for not allowing users to use unsmoothed fields for verification?

**Comment by JohnHalleyGotway** [comment]
Hi @jwraby, I see you have a question about the `interp` option in the Grid-Stat configuration file. You want to make sure that no smoothing is actually being performed on the data. And that is indeed the case.

Here's the [default interp setting ](https://github.com/dtcenter/MET/blob/5ffa5f15d9a8ca359a8888b25c8c12a1af4e4780/data/config/GridStatConfig_default#L145)from the default Grid-Stat config file:
```
interp = {
   field      = BOTH;
   vld_thresh = 1.0;
   shape      = SQUARE;

   type = [
      {
         method = NEAREST;
         width  = 1;
      }
   ];
}
```
As you can see on [this line of the smooth_field()](https://github.com/dtcenter/MET/blob/5ffa5f15d9a8ca359a8888b25c8c12a1af4e4780/src/basic/vx_util/data_plane_util.cc#L109) function, calling it with nearest neighbor with a width of 1 simply returns and does no actual work.

I can see how this it may be more intuitive to support a `field = NONE` setting, but the good news is that it is already behaving as you'd expect.

Do you think it'd be worthwhile to add support for `field = NONE` and update the default Grid-Stat configuration file to use that?

**Reply by jwraby** [reply]
Thanks for your reply, John. What I was seeing in the plot rendered from the Grid-Stat (python embedding) output ncpairs file was a very smoothed field of 2m TMP gridded obs produced by a python interpolation script which I had configured to generate a grid with 2.5 km spacing and had corresponding settings in the attrs dictionary (d_km=2.5, nx=79, ny=79) for the domain I'm working with. As a test, I ran plot_data_plane using python embedding for the interpolation script and had it produce a plot of the gridded obs produced by the python script and that plot was not as smooth-looking as expected. The python script independently generates its own plots of the gridded obs and that plot was identical to the plot generated by plot_data_plane. The plot rendered from the ncpairs file looks way too smooth compared to the other two plots which is why I asked the question about how to run G-S configured so that no smoothing occurs. Based on your response, I looked at my Grid-Stat config file and confirmed that I had the default interp setting, so I can only conclude now, that maybe something went awry when I ran Grid-Stat or plot_data_plane. I will re-check the settings in the python script and re-run the above experiment and let you know what the outcome is.

**Reply by jwraby** [reply]
I re-ran the experiment after confirming the settings in the python interpolation script for the 2.5 km output grid and the settings in the attrs dictionary which correspond to the 2.5 km grid and array dimensions. I'm pretty sure that the Grid-Stat and plot_data_plane commands were specifying the correct python script for the python embedding. The output matches what I was seeing as described above. The following files are attached:
2maglTMP_anal_test5_10APR2023_2.5km_grid.png (output of plot_data_plane)--> looks as expected
2.5km_TMP_12Z_12NOV2016_test5.png (output of Grid-Stat ncpairs file)--> looks too smooth
Figure_1_2pt5km_grid_10APR2023_test5.png (plot produced by python script)--> looks as expected

![Figure_1_2pt5km_grid_10APR2023_test5](https://user-images.githubusercontent.com/87020191/230991850-1e22ba1a-e84f-4fe9-b5f1-77633ff9ea9c.png)
![2 5km_TMP_12Z_12NOV2016_test5](https://user-images.githubusercontent.com/87020191/230991963-38c27302-ef14-4466-9c79-a38510b66a11.png)
![2maglTMP_anal_test5_10APR2023_2 5km_grid](https://user-images.githubusercontent.com/87020191/230992385-62ca1ff3-4ad7-43a8-9555-aefbe6355045.png)

I looked at the metadata of the ncpairs file and the dimension size (79,79) of teh 2m TMP field matches that of the grid specs in the attrs dictionary. Not sure what else to look at to understand why the plot of the 2m TMP field should look so different from the other two plots.

**Reply by JohnHalleyGotway** [reply]
@jwraby I'm just not seeing the differences that you describe in these plots. The overall contours of the data are the same in all 3 plots. The last one is certainly shown with more contours, but that's just a plotting option. It has nothing to do with the underlying data.

I can tell you that by default `plot_data_plane` rescales the plotting range to be the actual min/max values of the data being plotted. By convention, this rescaling is only done when the colortable used ranges from 0 to 1. Checking the colorbar extremes, I see that this data ranges from 276.988 to 282.362. And that overlaps with the range used for the first plot. To make the plots look more similar, you'd need to control the plotting range and the colortable used. A colortable can be coarse with only a handful of colors or very fine with many hundreds of colors.

You can actually test this with `plot_data_plane` using the `-plot_range` and `-color_table` command line options. You could play around with those if you'd like. MET's [default color table](https://github.com/dtcenter/MET/blob/main_v11.0/data/colortables/met_default.ctable) contains 41 colors. You could try using a [more coarse](https://github.com/dtcenter/MET/blob/main_v11.0/data/colortables/mode_raw.ctable) one with only 21 colors or a [more detailed](https://github.com/dtcenter/MET/blob/main_v11.0/data/colortables/NCL_colortables/rainbow.ctable) one with 188 colors. The number of contours and corresponding colors will change but the overall shape of those contours will remain consistent.

So I really don't see a problem here. I strongly suspect the underlying data is the same and the differences you're seeing are just based on the plotting range and colortables used by each plotting tool.

For extra confidence, you could see if each plotting tool can report the min/max value being plotted and then just make sure the shape of the data remains consistent across the output images.

**Reply by jwraby** [reply]
Thanks for sharing your insight into the impacts that plot settings can make on the appearance of the data. Those hints on adjusting the plot_range and color_table are helpful. I was looking more specifically at the the jagged appearance of the contours themselves in the plots from plot_data_plane and from the python interpolation script. From other experiments I've run, when I reduce the grid spacing from 2.5 km down to 0.5 km keeping the size of the domain constant, that jagged appearance disappears in the plots from plot_data_plane and the python script without any changes in the plot settings. This is an artifact of the "smoothing" achieved by increasing the number of grid points in the array from (79,79) to (393,393) in the same domain. As I increase the spacing, the more jagged the contours become. When you approach a spacing where you only have a few grid points in the array, say (9,9), the appearance of the contours becomes very blocky (extreme jaggedness). This is what I'm concerned about. 

Since the question I have is regarding the plot of the 2m TMP field rendered from the ncpairs output, I need to go back and  check/adjust the plot settings being used by the Unidata IDV which is what I used to generate the attached plot (2.5km_TMP_12Z_12NOV2016_test5.png). The metadata from the ncpairs file which is displayed in IDV shows the array with the correct dimensions (79,79) which is encouraging. Better yet would be to use plot_data_plane to plot that field from the ncpairs file. That way I'm not comparing apples and oranges in terms of plotting software. I will do this plot first so I can compare the two plots both rendered by plot_data_plane.

**Reply by jwraby** [reply]
OK. Ran into a problem when running plot_data_plane. I specified the input as the Grid-Stat ncpairs output file, but the name of the field contains white space:
ERROR  : VarInfo::set_magic() -> embedded whitespace found in the name "2mAGL_TMP_ANAL at Surface" or level "(*,*)" string. I obtained the long_name using ncdump on the ncpairs file. 

The field name originates from the attrs dictionary contained in the python interpolation script as follows:
attrs = {

   'valid':     '20161112_120000',
   'init':      '20161112_000000',
   'lead':      '000000',
   'accum':     '000000',

   'name':      'Z2TMP',
   'long_name': '2mAGL_TMP_ANAL',
   'level':     'Surface',
   'units':     'kelvin',

 'grid': {
      'type': 'Lambert Conformal',
      'hemisphere': 'N',
      'name': 'lambert',
      'scale_lat_1': 25.0,
      'scale_lat_2': 25.0,
      'lat_pin': 31.41,
      'lon_pin': -107.56,
      'x_pin': 0.0,
      'y_pin': 0.0,
      'lon_orient': -106.0,
      'd_km': 2.5,
      'r_km': 6367.47,
      'nx': 79,
      'ny': 79,
    }

When I ran Grid-Stat using python embedding, the ncpairs file was generated with the long_name containing the whitespace, but I'm not sure why this happened. Can you see the issue in the info I put in the above dictionary? Thanks.

**Reply by JohnHalleyGotway** [reply]
John, I don't see anything immediately obvious. You could send a sample data file along with the command line you ran to produce the error message. If I can replicate the runtime error locally, I should be able to figure out what's going on.

**Reply by jwraby** [reply]
Belay my last reply on the issue with field name containing whitespace. I had an operator error with using field names from the wrong ncpairs file. When I used the correct ncpairs field name it worked and the plot was generated by plot_data_plane. Furthermore, this plot shows the expected jagged appearance of the contours for the grid spacing of 2.5 km and (79,79) array dimensions! This means pretty conclusively, that the appearance of the plot generated by IDV was misleading me about the cause of the problem. Like you were suggesting, it appears that the plot settings are indeed causing the issue. Particularly, the settings in IDV need to be revised so that the expected jaggedness can be resolved. Thanks for your help on this!

**Reply by jwraby** [reply, marked as answer]
I concur with your recommendation to add support for field = NONE and update the default Grid-Stat configuration file to use that? It wasn't obvious to me that there was the option to perform verification on unsmoothed fields from the Grid-Stat config file nor from the User's Guide. Thanks for considering this enhancement.

**Reply by JohnHalleyGotway** [reply]
@jwraby, FYI, I wrote up this issue (dtcenter/MET#2510) to capture this requested change to the Grid-Stat config file. Not sure how it'll slot into our competing development priorities, but I just wanted to give you a heads up. I'll lock this discussion to prevent future posts. But feel free to comment directly on dtcenter/MET#2510 if you have any anything else to share about that potential change.
